forward_system/src/run/tracing_impl.rs

```diff
@@ -0,0 +1,338 @@
+    }
+
+    fn caller(&self) -> Address {
+        self.0.external_call.caller.to_be_bytes().into()
+    }
+
+    fn callee(&self) -> Address {
+        self.0.external_call.callee.to_be_bytes().into()
+    }
+
+    fn modifier(&self) -> zksync_os_interface::tracing::CallModifier {
+        match self.0.external_call.modifier {
+            CallModifier::NoModifier => zksync_os_interface::tracing::CallModifier::NoModifier,
+            CallModifier::Constructor => zksync_os_interface::tracing::CallModifier::Constructor,
+            CallModifier::Delegate => zksync_os_interface::tracing::CallModifier::Delegate,
+            CallModifier::Static => zksync_os_interface::tracing::CallModifier::Static,
+            CallModifier::DelegateStatic => {
+                zksync_os_interface::tracing::CallModifier::DelegateStatic
+            }
+            CallModifier::ZKVMSystem => zksync_os_interface::tracing::CallModifier::ZKVMSystem,
+            CallModifier::ZKVMSystemStatic => {
+                zksync_os_interface::tracing::CallModifier::ZKVMSystemStatic
+            }
+            CallModifier::EVMCallcode => zksync_os_interface::tracing::CallModifier::EVMCallcode,
+            CallModifier::EVMCallcodeStatic => {
+                zksync_os_interface::tracing::CallModifier::EVMCallcodeStatic
+            }
+        }
+    }
+
+    fn input(&self) -> &[u8] {
+        self.0.external_call.input
+    }
+
+    fn nominal_token_value(&self) -> U256 {
+        self.0.external_call.nominal_token_value
+    }
+}
+
+impl<'a, S: EthereumLikeTypes, T: EvmFrameInterface<S>>
+    zksync_os_interface::tracing::EvmFrameInterface for EvmFrameInterfaceWrapped<'a, S, T>
+{
+    fn instruction_pointer(&self) -> usize {
+        self.inner.instruction_pointer()
+    }
+
+    fn resources(&self) -> EvmResources {
+        let resouces = self.inner.resources();
+        EvmResources {
+            ergs: resouces.ergs().0,
+            native: resouces.native().as_u64(),
+        }
+    }
+
+    fn caller(&self) -> Address {
+        self.inner.caller().to_be_bytes().into()
+    }
+
+    fn address(&self) -> Address {
+        self.inner.address().to_be_bytes().into()
+    }
+
+    fn calldata(&self) -> &[u8] {
+        self.inner.calldata()
+    }
+
+    fn return_data(&self) -> &[u8] {
+        self.inner.return_data()
+    }
+
+    fn heap(&self) -> &[u8] {
+        self.inner.heap()
+    }
+
+    fn bytecode(&self) -> &[u8] {
+        self.inner.bytecode()
+    }
+
+    fn call_value(&self) -> &U256 {
+        self.inner.call_value()
+    }
+
+    fn refund_counter(&self) -> u32 {
+        self.inner.refund_counter()
+    }
+
+    fn is_static(&self) -> bool {
+        self.inner.is_static()
+    }
+
+    fn is_constructor(&self) -> bool {
+        self.inner.is_constructor()
+    }
+}
+
+fn convert_error(error: &EvmError) -> zksync_os_interface::error::EvmError {
+    match error {
+        EvmError::Revert => zksync_os_interface::error::EvmError::Revert,
```

Comment:
Imo it's better to use zksync_os_interface::error::EvmError inside of ZKsync OS instead

Reply:
Unfortunately this doesn't work due to std somewhere in deps etc.

Reply:
Probably we should move EVM errors to a separate crate, wdyt?

Reply:
I can make the whole crate `no_std` pretty easily. In fact I have already done it once during the first phase of integration. Using it across the code had a significant impact on benchmarks as you might remember. But if we limit the non-forward usage to just errors then it should be fine.

Sorry this is dragging on a bit, I got distracted by other work but I am hoping to bring this to a finish today/tomorrow.

Reply:
Actually it was too much hassle (forgot that it needed all that stuff with portable atomic and single-threaded CAS) so I crated another crate as you suggested. PTAL